src/docker/dockerUtils.ts

```diff
@@ -0,0 +1,110 @@
+    // https://docs.docker.com/engine/reference/commandline/tag/#extended-description
+
+    return name.toLowerCase().replace(/[^a-z0-9._-]/g, '-');
+}
+
+async function getDefaultImageName(cwd: string, imagePrefix?: string): Promise<string> {
+    const name = sanitiseTag(path.basename(cwd));
+    const version = await findVersion(cwd);
+    let image = `${name}:${version}`;
+    if (imagePrefix) {
+        image = imagePrefix + "/" + image;
+    }
+    return image;
+}
+
+async function findVersion(cwd: string): Promise<string> {
+    const shellOpts = Object.assign({ }, shell.execOpts(), { cwd });
+    const shellResult = await shell.execCore('git describe --always --dirty', shellOpts);
+    return shellResult.code !== 0 ? "latest" : shellResult.stdout.trim();
+}
+
+async function buildDockerImage(dockerClient: DockerClient, image: string, shellOpts: any): Promise<void> {
+    // Build docker image.
+    const buildResult = await shell.execCore(`${dockerClient} build -t ${image} .`, shellOpts);
+    if (buildResult.code !== 0) {
+        throw new Error(`Image build failed: ${buildResult.stderr}`);
+    }
+    kubeChannel.showOutput(image + ' built.');
+}
+
+async function pushDockerImage(dockerClient: string, image: string, shellOpts: any): Promise<void> {
+    // Push docker image.
+    const pushResult = await shell.execCore(`${dockerClient} push ${image}`, shellOpts);
+    if (pushResult.code !== 0) {
+        throw new Error(`Image push failed: ${pushResult.stderr}`);
+    }
+    kubeChannel.showOutput(image + ' pushed.');
+}
+
+/**
+ * When using the command "minikube docker-env" to get the local kubernetes docker env, it needs run with the admin privilege.
+ * To workaround this, this function will try to resolve the equivalent docker env from kubeconfig instead.
+ */
+export async function resolveDockerEnv(kubectl: Kubectl): Promise<{}> {
+    const dockerEnv = {};
+    dockerEnv["DOCKER_API_VERSION"] = await dockerApiVersion();
+    const currentCluster = await getCurrentClusterConfig(kubectl);
+    if (!currentCluster || !currentCluster.server || !currentCluster.certificateAuthority) {
+        return {};
+    }
+    
+    if (/^https/.test(currentCluster.server)) {
+        dockerEnv["DOCKER_TLS_VERIFY"] = 1;
+    }
+    const serverUrl = url.parse(currentCluster.server);
+    dockerEnv["DOCKER_HOST"] = `tcp://${serverUrl.hostname}:2376`;
+    const certDir = path.dirname(currentCluster.certificateAuthority);
+    if (fs.existsSync(path.join(certDir, "certs"))) {
+        dockerEnv["DOCKER_CERT_PATH"] = path.join(certDir, "certs");
+    } else {
+        dockerEnv["DOCKER_CERT_PATH"] = certDir;
+    }
+    return dockerEnv;
+}
+
+async function dockerApiVersion(): Promise<string> {
+    const versionResult = await shell.exec(`docker version --format "{{.Client.APIVersion}}"`);
+    if (versionResult.code === 0) {
+        return versionResult.stdout.trim();
+    }
+    return "1.23";

```

Comment:
Where does this magic string come from?  Can we make it a named constant?

Reply:
This is the current docker version, just want to add a default version. Will assign it to a named constant.